src/internal/operators/window.ts

```diff
@@ -50,77 +49,37 @@ import { SimpleOuterSubscriber, innerSubscribe, SimpleInnerSubscriber } from '..
  * @name window
  */
 export function window<T>(windowBoundaries: Observable<any>): OperatorFunction<T, Observable<T>> {
-  return function windowOperatorFunction(source: Observable<T>) {
-    return lift(source, new WindowOperator(windowBoundaries));
-  };
-}
-
-class WindowOperator<T> implements Operator<T, Observable<T>> {
-  constructor(private windowBoundaries: Observable<any>) {}
-
-  call(subscriber: Subscriber<Observable<T>>, source: any): any {
-    const windowSubscriber = new WindowSubscriber(subscriber);
-    const sourceSubscription = source.subscribe(windowSubscriber);
-    if (!sourceSubscription.closed) {
-      windowSubscriber.add(innerSubscribe(this.windowBoundaries, new SimpleInnerSubscriber(windowSubscriber)));
-    }
-    return sourceSubscription;
-  }
-}
-
-/**
- * We need this JSDoc comment for affecting ESDoc.
- * @ignore
- * @extends {Ignored}
- */
-class WindowSubscriber<T> extends SimpleOuterSubscriber<T, any> {
-  private window: Subject<T> = new Subject<T>();
-
-  constructor(destination: Subscriber<Observable<T>>) {
-    super(destination);
-    destination.next(this.window);
-  }
-
-  notifyNext(): void {
-    this.openWindow();
-  }
-
-  notifyError(error: any): void {
-    this._error(error);
-  }
-
-  notifyComplete(): void {
-    this._complete();
-  }
-
-  protected _next(value: T): void {
-    this.window.next(value);
-  }
-
-  protected _error(err: any): void {
-    this.window.error(err);
-    this.destination.error(err);
-  }
+  return (source: Observable<T>) =>
+    lift(source, function (this: Subscriber<Observable<T>>, source: Observable<T>) {
+      const subscriber = this;
+      let window = new Subject<T>();
 
-  protected _complete(): void {
-    this.window.complete();
-    this.destination.complete();
-  }
+      subscriber.next(window.asObservable());
 
-  unsubscribe() {
-    if (!this.closed) {
-      this.window = null!;
-      super.unsubscribe();
-    }
-  }
+      const windowSubscribe = (source: Observable<any>, next: (value: any) => void) =>
+        source.subscribe(
+          new OperatorSubscriber(
+            subscriber,
+            next,
+            (err: any) => {
+              window.error(err);
+              subscriber.error(err);
+            },
+            () => {
+              window.complete();
+              subscriber.complete();
+            },
+            () => {
+              window?.unsubscribe();
+              window = null!;
```

Comment:
issue: IMO, this usage means that `window` should be typed as `Subject<T> | null`. If optional chaining is needed hear, AFAICT, it will also be needed in the `error` and `complete` handlers and type current typing - as `Subject<T>` - is hiding what appears to be an error.

Reply:
you know... it just occurred to me that I might change the name of that variable too. Just so nothing bad happens where someone deletes `let`. (The operator itself should be changed as well, obviously, but we've never really agreed on what that should be)

Reply:
Okay, I've renamed the variable, and I've moved that teardown so it wasn't called twice (that way the conditional was unnecessary)